source/MaterialXTest/MaterialXCore/Node.cpp

```diff
@@ -675,7 +677,64 @@ TEST_CASE("Organization", "[nodegraph]")
     CHECK(nodeGraph->getBackdrops().empty());
 }
 
-TEST_CASE("Node Definition Creation", "[nodedef]")
+void testFunctionalNodeDef()
+{
+    mx::FileSearchPath searchPath = mx::getDefaultDataSearchPath();
+    mx::DocumentPtr doc = mx::createDocument();
+    mx::readFromXmlFile(doc, "resources/Materials/TestSuite/stdlib/definition/functional_nodedef.mtlx", searchPath);
+
+    std::vector<mx::NodeDefPtr> nodedefs = doc->getNodeDefs();
+    for (mx::NodeDefPtr nodeDef : nodedefs)
+    {
+        std::string nodeDefName = nodeDef->getName();
+        std::string nodeGraphName = "NG_" + nodeDefName.substr(3); // Remove the 'ND_' prefix
+
+        mx::InterfaceElementPtr implementation = nodeDef->getImplementation();
+        REQUIRE(implementation != nullptr);
+        std::string msg = "Testing NodeDef: " + nodeDefName + " with implementation: " +  implementation->getName();
+        INFO(msg);
+        mx::NodeGraphPtr functionalNodeGraph = implementation->asA<mx::NodeGraph>();
+        REQUIRE(functionalNodeGraph != nullptr);
+        if (functionalNodeGraph)
+        {
+            // Test that the child nodegraph is found via implementation search
+            REQUIRE(functionalNodeGraph->getName() == nodeGraphName);
+
+            // Test that this is actually a child nodegraph of the NodeDef
+            std::vector<mx::NodeGraphPtr> childNodeGraphs = nodeDef->getChildrenOfType<mx::NodeGraph>();
+            for (mx::NodeGraphPtr childNodeGraph : childNodeGraphs)
+            {
+                if (childNodeGraph->getName() == nodeGraphName)
+                {
+                    // Test that the child nodegraph is the functional graph
+                    REQUIRE(childNodeGraph == functionalNodeGraph);
+                }
+            }
+        }
+
+        std::string referenceGraphName = nodeGraphName + "_reference";
+        mx::NodeGraphPtr referenceNodeGraph = doc->getNodeGraph(referenceGraphName);
+        REQUIRE(referenceNodeGraph != nullptr);
+        if (referenceNodeGraph)
+        {
+            referenceNodeGraph->setNodeDefString(nodeDefName);
```

Comment:
Superscede the child nodegraph impl with one that is not a child. As all changes cause a nodedef->impl cache rebuild this allows for this dynamic behaviour.